src/solace_agent_mesh/agent/tools/deep_research_tools.py

```diff
@@ -0,0 +1,1761 @@
+            sources=list(self.citations.values()),
+            metadata={
+                "queries": self.queries  # Include query breakdown for timeline
+            }
+        )
+
+
+async def _send_research_progress(
+    message: str,
+    tool_context: ToolContext,
+    phase: str = "",
+    progress_percentage: int = 0,
+    current_iteration: int = 0,
+    total_iterations: int = 0,
+    sources_found: int = 0,
+    current_query: str = "",
+    fetching_urls: Optional[List[Dict[str, str]]] = None,
+    elapsed_seconds: int = 0,
+    max_runtime_seconds: int = 0
+) -> None:
+    """Send research progress update to frontend via SSE with structured data"""
+    log_identifier = "[DeepResearch:Progress]"
+    
+    try:
+        # Get a2a context from tool context state
+        a2a_context = tool_context.state.get("a2a_context")
+        if not a2a_context:
+            log.warning("%s No a2a_context found, cannot send progress update", log_identifier)
+            return
+
+        # Get the host component from invocation context
+        invocation_context = getattr(tool_context, '_invocation_context', None)
+        if not invocation_context:
+            log.warning("%s No invocation context found", log_identifier)
+            return
+            
+        agent = getattr(invocation_context, 'agent', None)
+        if not agent:
+            log.warning("%s No agent found in invocation context", log_identifier)
+            return
+            
+        host_component = getattr(agent, 'host_component', None)
+        if not host_component:
+            log.warning("%s No host component found on agent", log_identifier)
+            return
+
+        log.info("%s Sending progress: %s", log_identifier, message)
+
+        # Use structured DeepResearchProgressData if phase is provided, otherwise simple text
+        from ...common.data_parts import DeepResearchProgressData, AgentProgressUpdateData
+        
+        if phase:
+            # Send structured progress data for UI visualization
+            progress_data = DeepResearchProgressData(
+                phase=phase,
+                status_text=message,
+                progress_percentage=progress_percentage,
+                current_iteration=current_iteration,
+                total_iterations=total_iterations,
+                sources_found=sources_found,
+                current_query=current_query,
+                fetching_urls=fetching_urls or [],
+                elapsed_seconds=elapsed_seconds,
+                max_runtime_seconds=max_runtime_seconds
+            )
+        else:
+            # Fallback to simple text progress
+            progress_data = AgentProgressUpdateData(status_text=message)
+        
+        logical_task_id = a2a_context.get("logical_task_id")
+        context_id = a2a_context.get("contextId")
+        
+        # Create status update event using the standard data signal pattern
+        status_update_event = a2a.create_data_signal_event(
```

Comment:
Hmm... this really makes me think we need to get a better tool framework that can provide these services to tools rather than having them call publishing routines directly. Way too much boilerplate here. Ideally, we could have a way to define status updates for the tool in a generic way and then be able to just call a higher level function with status updates... 

I wouldn't hold up this feature for that, but I think we need to build that before we do more of this